src/Microsoft.ML.PipelineInference/PipelinePattern.cs

```diff
@@ -183,57 +204,80 @@ public Experiment CreateTrainTestExperiment(IDataView trainData, IDataView testD
+        public void RunTrainTestExperiment(IDataView trainData, IDataView testData,
+            AutoInference.SupportedMetric metric, MacroUtils.TrainerKinds trainerKind, out double testMetricValue,
+            out double trainMetricValue)
         {
-            var experiment = CreateTrainTestExperiment(trainData, testData, trainerKind, out var trainTestOutput);
+            var experiment = CreateTrainTestExperiment(trainData, testData, trainerKind, true, out var trainTestOutput);
             experiment.Run();
-            var dataOut = experiment.GetOutput(trainTestOutput.OverallMetrics);
-            var schema = dataOut.Schema;
-            schema.TryGetColumnIndex(metric.Name, out var metricCol);
 
-            using (var cursor = dataOut.GetRowCursor(col => col == metricCol))
-            {
-                var getter = cursor.GetGetter<double>(metricCol);
-                double metricValue = 0;
-                cursor.MoveNext();
-                getter(ref metricValue);
-                return metricValue;
-            }
+            var dataOut = experiment.GetOutput(trainTestOutput.OverallMetrics);
+            var dataOutTraining = experiment.GetOutput(trainTestOutput.TrainingOverallMetrics);
+            testMetricValue = AutoMlUtils.ExtractValueFromIDV(_env, dataOut, metric.Name);
+            trainMetricValue = AutoMlUtils.ExtractValueFromIDV(_env, dataOutTraining, metric.Name);
         }
 
-        public static PipelineResultRow[] ExtractResults(IHostEnvironment env, IDataView data, string graphColName, string metricColName, string idColName)
+        public static PipelineResultRow[] ExtractResults(IHostEnvironment env, IDataView data,
+            string graphColName, string metricColName, string idColName, string trainingMetricColName,
+            string firstInputColName, string predictorModelColName)
         {
             var results = new List<PipelineResultRow>();
             var schema = data.Schema;
             if (!schema.TryGetColumnIndex(graphColName, out var graphCol))
-                throw env.ExceptNotSupp($"Column name {graphColName} not found");
+                throw env.ExceptParam(graphColName, $"Column name {graphColName} not found");
             if (!schema.TryGetColumnIndex(metricColName, out var metricCol))
-                throw env.ExceptNotSupp($"Column name {metricColName} not found");
+                throw env.ExceptParam(metricColName, $"Column name {metricColName} not found");
+            if (!schema.TryGetColumnIndex(trainingMetricColName, out var trainingMetricCol))
+                throw env.ExceptParam(trainingMetricColName, $"Column name {trainingMetricColName} not found");
             if (!schema.TryGetColumnIndex(idColName, out var pipelineIdCol))
-                throw env.ExceptNotSupp($"Column name {idColName} not found");
+                throw env.ExceptParam(idColName, $"Column name {idColName} not found");
```

Comment:
>idColName [](start = 38, length = 9)

Note that this ought to be the parameter name... so `nameof(idColName)`, not just the value that the parameter holds.

Reply:
Same for others here.

---
In reply to: [190064643](https://github.com/dotnet/machinelearning/pull/152#discussion_r190064643) [](ancestors = 190064643)

Reply:
Can you clarify a bit, I'm not sure I'm understanding how CheckParam works in reference to your comment. At this part of the code we need to check if a certain column name (stored in idColName, such as "FooCol") is present in the schema. If it is not (as indicated by the failure to get the index column) we need to throw an exception to let the user know that the column name it is searching for, FooCol, is not present. We don't want to tell them that "idColName" is not present. (This is how I interpret your comment, which doesn't seem right.) Thanks!